core/src/apps/cardano/auxiliary_data.py

```diff
@@ -0,0 +1,252 @@
+from trezor.messages import (
+    CardanoAddressType,
+    CardanoAuxiliaryDataType,
+    CardanoMetadataType,
+)
+
+from apps.common import cbor
+
+from ..common.seed import remove_ed25519_prefix
+from .address import derive_address_bytes, derive_human_readable_address
+from .helpers import INVALID_AUXILIARY_DATA, bech32
+from .helpers.bech32 import HRP_JORMUN_PUBLIC_KEY
+from .helpers.paths import SCHEMA_STAKING_ANY_ACCOUNT
+from .layout import confirm_catalyst_registration, show_auxiliary_data_hash
+
+if False:
+    from typing import Union
+    from trezor import wire
+
+    from trezor.messages.CardanoCatalystRegistrationParametersType import (
+        CardanoCatalystRegistrationParametersType,
+    )
+    from trezor.messages.CardanoTxAuxiliaryDataType import CardanoTxAuxiliaryDataType
+
+    CatalystRegistrationPayload = dict[int, Union[bytes, int]]
+    CatalystRegistrationSignature = dict[int, bytes]
+    CatalystRegistration = dict[
+        int, Union[CatalystRegistrationPayload, CatalystRegistrationSignature]
+    ]
+
+    from . import seed
+
+AUXILIARY_DATA_HASH_SIZE = 32
+PUBLIC_KEY_LENGTH = 32
+CATALYST_REGISTRATION_HASH_SIZE = 32
+CATALYST_REGISTRATION_KEY = 61284
+CATALYST_REGISTRATION_SIGNATURE_KEY = 61285
+
+
+def validate_auxiliary_data(
+    auxiliary_data: CardanoTxAuxiliaryDataType | None,
+) -> None:
+    if not auxiliary_data:
+        return
+
+    if auxiliary_data.type == CardanoAuxiliaryDataType.BLOB:
+        if not auxiliary_data.blob:
+            raise INVALID_AUXILIARY_DATA
+
+        _validate_auxiliary_data_blob(auxiliary_data.blob)
+    elif auxiliary_data.type == CardanoAuxiliaryDataType.TUPLE:
+        # currently we only support CATALYST_REGISTRATION
+        if (
+            auxiliary_data.blob
+            or not auxiliary_data.metadata
+            or auxiliary_data.metadata.type != CardanoMetadataType.CATALYST_REGISTRATION
+        ):
+            raise INVALID_AUXILIARY_DATA
+
+        _validate_catalyst_registration_parameters(
+            auxiliary_data.metadata.catalyst_registration_parameters
+        )
+    else:
+        raise INVALID_AUXILIARY_DATA
+
+
+def _validate_auxiliary_data_blob(auxiliary_data_blob: bytes) -> None:
+    try:
+        # this also raises an error if there's some data remaining
+        cbor.decode(auxiliary_data_blob)
+    except Exception:
+        raise INVALID_AUXILIARY_DATA
+
+
+def _validate_catalyst_registration_parameters(
+    catalyst_registration_parameters: CardanoCatalystRegistrationParametersType | None,
+) -> None:
+    if not catalyst_registration_parameters:
+        raise INVALID_AUXILIARY_DATA
+
+    if len(catalyst_registration_parameters.voting_public_key) != PUBLIC_KEY_LENGTH:
+        raise INVALID_AUXILIARY_DATA
+
+    if not SCHEMA_STAKING_ANY_ACCOUNT.match(
+        catalyst_registration_parameters.staking_path
+    ):
+        raise INVALID_AUXILIARY_DATA
+
+    address_parameters = catalyst_registration_parameters.reward_address_parameters
```

Comment:
More validation is actually happening as a part of `derive_address_bytes`. It is a bit weird and should probably be refactored since in other places first validation happens and then the work. I don't think it should be done as part of this PR though, so I've created an issue for that - https://github.com/vacuumlabs/trezor-firmware/issues/59.

For now, I've also added a call to `derive_address_bytes` as part of the validation as is done when validating outputs during sign_tx. https://github.com/vacuumlabs/trezor-firmware/pull/57/commits/de6ce66e59dc846769a71ec5ce9fa3e0583d45a1#diff-449dbb3cc07455194600fd241faad462fb0b7417008b0b99fef57c4c7966745bR120